packages/EasyCore/tests/Bridge/Symfony/Serializer/TrimStringsNormalizerTest.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+declare(strict_types=1);
+
+namespace EonX\EasyCore\Tests\Bridge\Symfony\Serializer;
+
+use EonX\EasyCore\Bridge\Symfony\Serializer\TrimStringsNormalizer;
+use EonX\EasyCore\Helpers\StringsTrimmerInterface;
+use EonX\EasyCore\Tests\Bridge\Symfony\AbstractSymfonyTestCase;
+use EonX\EasyCore\Tests\Bridge\Symfony\Stubs\NormalizerStub;
+use Mockery\MockInterface;
+use stdClass;
+
+/**
+ * @covers \EonX\EasyCore\Bridge\Symfony\Serializer\TrimStringsNormalizer
+ *
+ * @internal
+ */
+final class TrimStringsNormalizerTest extends AbstractSymfonyTestCase
+{
+    /**
+     * @return mixed[]
+     *
+     * @see testDenormalizeSucceedsWithTrimValue
+     */
+    public function provideDataForDenormalize(): array
+    {
+        $object = new stdClass();
+
+        return [
+            'data is string' => [
+                '',
+                '',
+            ],
+            'data is array' => [
+                [],
+                [],
+            ],
+            'data is bool' => [
+                true,
+                true,
+            ],
+            'data is numeric' => [
+                123,
+                123,
+            ],
+            'data is object' => [
+                $object,
+                $object,
+            ],
+        ];
+    }
+
+    /**
+     * @param mixed $data
+     * @param mixed $expectedResult
+     *
+     * @dataProvider provideDataForDenormalize
+     *
+     * @throws \Symfony\Component\Serializer\Exception\ExceptionInterface
+     */
+    public function testDenormalizeSucceedsWithTrimValue($data, $expectedResult): void
+    {
+        $type = 'no-matter';
+        $format = null;
+        $context = [];
+        $except = ['some-key'];
+        $expectation = static function (MockInterface $mock): void {
+            $mock->shouldNotReceive('trim');
+        };
+        if (\is_string($data) || \is_array($data)) {
+            $expectation = static function (MockInterface $mock) use ($data, $except, $expectedResult): void {
+                $mock->shouldReceive('trim')->once()->with($data, $except)->andReturn($expectedResult);
+            };
+        }
+        /** @var \EonX\EasyCore\Helpers\StringsTrimmerInterface $trimmer */
+        $trimmer = $this->mock(
+            StringsTrimmerInterface::class,
+            $expectation
+        );
+        $decorated = new NormalizerStub();
+        $denormalizer = new TrimStringsNormalizer($decorated, $trimmer, $except);
```

Comment:
I don't think that this test is enough, we should use the real serializer instead of a stub because I believe our normalizer would break the application as it would create an infinite loop.
Had the issue in other projects where we had to add flag to the context not to process data twice.

Reply:
@natepage We have already used a similar approach (decorating the API service) in BMS. And there were no problems with an infinite loop or data reprocessing. I also checked the work within the EWSP, no problems using the service.
@roman-eonx @egor-dev What do you think?

Reply:
Let's introduce a flag and add integration tests for the middleware and normalizer. Probably we don't even need unit tests (for these two classes) after that.

Reply:
Ok, let's add a flag, just in case.

Reply:
@natepage @roman-eonx @egor-dev For correct testing, we cannot use either unit or integration tests. The occurrence of an infinite loop can be checked only in a full-fledged application where a real container with services and their dependencies is used. This means that we can only rely on a functional test, which, within the framework of a real application, for example, EWMP, will allow us to make sure that this code works.